uw_oidc/id_token.py

```diff
@@ -1,82 +1,94 @@
+import logging
+from calendar import timegm
+from datetime import datetime, timedelta, timezone
 from django.conf import settings
-from jwt import decode
+from jwt import decode, get_unverified_header
 from jwt.exceptions import PyJWTError, InvalidSignatureError
-from uw_oidc.exceptions import InvalidTokenError
-from restclients_core.dao import DAO
-from restclients_core.exceptions import DataFailureException
-from os.path import abspath, dirname
-import os
-import json
+from uw_oidc.exceptions import (
+    InvalidTokenError, InvalidTokenHeader, NoMatchingPublicKey)
+from uw_oidc.jwks import UW_JWKS
 
-
-class UWIDP_DAO(DAO):
-    def service_name(self):
-        return 'uw_idp'
-
-    def service_mock_paths(self):
-        return [abspath(os.path.join(dirname(__file__), 'resources'))]
-
-    def delete_cache_key(self, url):
-        cache = self.get_cache()
-        cache_key = cache._get_key(self.service_name(), url)
-        cache.client.delete(cache_key)
+logger = logging.getLogger(__name__)
 
 
 class UWIdPToken(object):
+    JWKS_CLIENT = UW_JWKS()
     JWT_OPTIONS = {
-        'require_exp': True, 'require_iat': True, 'verify_signature': True,
-        'verify_iat': True, 'verify_exp': True, 'verify_iss': True,
-        'verify_aud': True,
+        'require_exp': True, 'require_iat': True,
+        'verify_signature': True, 'verify_iat': True, 'verify_exp': True,
+        'verify_iss': True, 'verify_aud': True
     }
-    SIGNING_ALGORITHMS = [
-        'RS256', 'RS384', 'RS512', 'HS256', 'HS384', 'HS512', 'ES256'
-    ]
-    KEY_URL = '/idp/profile/oidc/keyset'
-
-    def __init__(self, token):
-        self.token = token
+    # To avoid algorithm confusion attacks, always specify only
+    # the algoriths expected to use for token signature validation.
+    SIGNING_ALGORITHMS = ['RS256', 'RS384', 'RS512', 'ES256']
```

Comment:
The HS* algorithms are no longer supported (see https://idp-eval.u.washington.edu/.well-known/openid-configuration)